File: codes/java/239.滑动窗口最大值.java

class Solution {
	public int[] maxSlidingWindow(int[] nums, int k) {
		int length = nums.length;
		int[] result = new int[length - k + 1];
		Deque<Integer> deque = new LinkedList<>();
		for (int i = 0; i < length; i++) {
			if (i < k - 1) {
				while (!deque.isEmpty() && deque.peekLast() < nums[i]) {
					deque.pollLast();
				}
				deque.offerLast(nums[i]);
				continue;
			}
			while (!deque.isEmpty() && deque.peekLast() < nums[i]) {
				deque.pollLast();
			}
			deque.offerLast(nums[i]);
			result[i - k + 1] = deque.peekFirst();
			if (deque.peekFirst() == nums[i - k + 1]) {
				deque.pollFirst();
			}
		}
		return result;
	}
}

File: codes/python/239.滑动窗口最大值.py

import collections

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        length = len(nums)
        result = []
        deque = collections.deque()
        for i in range(length):
            if i < k - 1:
                while deque and deque[-1] < nums[i]:
                    deque.pop()
                deque.append(nums[i])
                continue
            while deque and deque[-1] < nums[i]:
                deque.pop()
            deque.append(nums[i])
            result.append(deque[0])
            if deque[0] == nums[i - k + 1]:
                deque.popleft()
        return result